modules/phase_field/src/userobjects/EBSDReader.C

```diff
@@ -244,123 +246,85 @@ EBSDReader::readFile()
-      quat_to_bin(_grain_num);
-
-  for (unsigned int i = 0; i < _grain_num; i++)
+
+  for (const auto i : make_range(_grain_num))
   {
     EBSDAvgData & a = _avg_data[i];
     EulerAngles & b = _avg_angles[i];
 
     if (a._n == 0)
       continue;
 
-    Real w_index, x_index, y_index, z_index;
-    Real w, x, y, z;
-    std::vector<std::tuple<double, Eigen::VectorXd>> eigen_vectors_and_values;
+    // creating a map to store the quaternion count for each bin index
+    std::map<std::tuple<int, int, int, int>, unsigned int> feature_weights;
 
-    // creating an empty map
-    feature_weights.push_back(
-        std::map<std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>,
-                 unsigned int>());
-
-    // looping through quaternions of each grain
-    for (unsigned int k = 0; k < quat[i].size(); k++)
+    // looping through all quaternions of the current grain
+    for (const auto & q : quat[i])
     {
-      std::tuple<unsigned int, unsigned int, unsigned int, unsigned int> bin;
-      w_index = int((quat[i][k].w() + 1) * 0.5 * _bins);
-      x_index = int((quat[i][k].x() + 1) * 0.5 * _bins);
-      y_index = int((quat[i][k].y() + 1) * 0.5 * _bins);
-      z_index = int((quat[i][k].z() + 1) * 0.5 * _bins);
-
-      bin = std::make_tuple(w_index, x_index, y_index, z_index);
-
-      // storing the bincorresponding to the quaternion quat[i][k]
-      quat_to_bin[i].push_back(bin);
-
-      // storing the bin name and its corresponding size value
-      auto j = feature_weights[i].find(bin);
-      if (j == feature_weights[i].end())
-        feature_weights[i].emplace_hint(j, bin, 1);
-      else
-        j->second++;
+      const auto bin = std::make_tuple<int, int, int, int>(std::floor(q.w() * 0.5 * _bins),
+                                                           std::floor(q.x() * 0.5 * _bins),
+                                                           std::floor(q.y() * 0.5 * _bins),
+                                                           std::floor(q.z() * 0.5 * _bins));
+      feature_weights[bin]++;
     }
 
     /**
      * Markley, F. Landis, Yang Cheng, John Lucas Crassidis, and Yaakov Oshman.
      * "Averaging quaternions." Journal of Guidance, Control, and Dynamics 30,
      * no. 4 (2007): 1193-1197.
      * A 4 by N matrix (Q) is constructed, where N is the number of quaternions.
-     * A weight matrix (W) is created. The eigen vector corresponding to the
-     * maximum eigen value of Q*W*Q' is the weighted average quaternion
-     **/
+     * A weight matrix (W) is created. The eigenvector corresponding to the
+     * maximum eigenvalue of Q*W*Q' is the weighted average quaternion
+     */
 
-    // creating a quaternion matrix
-    Eigen::MatrixXd quat_mat(4, quat[i].size());
-    // creating the weight matrix
-    Eigen::MatrixXd weight = Eigen::MatrixXd::Identity(quat[i].size(), quat[i].size());
```

Comment:
I can't see how the previous method was creating a 4xN matrix per your comment:

https://github.com/idaholab/moose/issues/15418#issuecomment-1014750495

Reply:
in line 304 `quat[i].size()` would be the number of datapoints for a given grain in the EBSD file. Those could be thousands.